### PR TITLE
chore: upgrade react-query

### PR DESCRIPTION
**Pull request by tihuan**
### Reviewers
**Functional:** 
@seve 

**Readability:** 
@frano-m @MillenniumFalconMechanic 

---

## Changes
- add
Upgrade `react-query` to v3!

Some breaking changes that affect us:

Source: https://react-query.tanstack.com/guides/migrating-to-react-query-3

1. useQuery:
    1. https://react-query.tanstack.com/guides/migrating-to-react-query-3#query-key-partspieces-are-no-longer-automatically-spread-to-the-query-function
    1. https://react-query.tanstack.com/guides/migrating-to-react-query-3#the-object-syntax-for-usequery-now-uses-a-collapsed-config
  
1. Mutation:
    1. https://react-query.tanstack.com/guides/migrating-to-react-query-3#usemutation-now-returns-an-object-instead-of-an-array
    1. https://react-query.tanstack.com/guides/migrating-to-react-query-3#mutationmutate-no-longer-return-a-promise. Use mutate for callback, and mutateAsync for await

1. QueryOptions.enabled needs a boolean value
    1. https://react-query.tanstack.com/guides/migrating-to-react-query-3#if-set-the-queryoptionsenabled-option-must-be-a-boolean-truefalse

1. TypeScript:
    1. QueryStatus: enums to string union
https://react-query.tanstack.com/guides/migrating-to-react-query-3#typescript

1. New Features:
    1. https://react-query.tanstack.com/guides/migrating-to-react-query-3#new-features

Live link:

https://thuang-upgrade-react-query-frontend.rdev.single-cell.czi.technology/datasets

## Definition of Done (from ticket)
Every feature should still work as expected!

Thank you 🙏 ✨

### CODE REVIEW

File: frontend/package.json [hunk 1]
```diff
@@ -57,7 +57,6 @@
     "playwright": "^1.13.1",
     "prettier": "^2.4.1",
     "prettier-plugin-organize-imports": "^2.3.4",
-    "react-query-devtools": "^2.6.3",
```
Comment (tihuan) [inline frontend/package.json:60]: devtools is now part of `react-query/devtools`

File: frontend/src/common/queries/auth.ts [hunk 2]
```diff
@@ -37,7 +36,7 @@ async function fetchUserInfo(
 const AUTH_TOKEN_CACHE_TIME_MS = 8 * 60 * 60 * 1000;
 
 export function useUserInfo(hasAuth: boolean) {
-  return useQuery([USE_USER_INFO, { hasAuth }], fetchUserInfo, {
+  return useQuery([USE_USER_INFO, { hasAuth }], () => fetchUserInfo(hasAuth), {
```
Comment (tihuan) [inline frontend/src/common/queries/auth.ts:39]: https://react-query.tanstack.com/guides/migrating-to-react-query-3#query-key-partspieces-are-no-longer-automatically-spread-to-the-query-function

File: frontend/src/common/queries/collections.ts [hunk 3]
```diff
@@ -153,12 +157,16 @@ function fetchCollection(allCollections: CollectionResponsesMap | undefined) {
...
+  id?: string;
+  visibility: VISIBILITY_TYPE;
+}): UseQueryResult<Collection | TombstonedCollection | null> {
   const { data: collections } = useCollections();
   const queryFn = fetchCollection(collections);
+
   return useQuery<Collection | TombstonedCollection | null>(
     [USE_COLLECTION, id, visibility, collections],
-    queryFn,
+    () => queryFn(id, visibility),
```
Comment (tihuan) [inline frontend/src/common/queries/collections.ts:169]: https://react-query.tanstack.com/guides/migrating-to-react-query-3#query-key-partspieces-are-no-longer-automatically-spread-to-the-query-function

File: frontend/src/common/queries/collections.ts [hunk 4]
```diff
@@ -182,16 +190,18 @@ export async function createCollection(payload: string): Promise<string> {
 }
 
 export function useCreateCollection() {
-  const queryCache = useQueryCache();
+  const queryClient = useQueryClient();
 
   return useMutation(createCollection, {
     onSuccess: () => {
-      queryCache.invalidateQueries(USE_COLLECTIONS);
+      queryClient.invalidateQueries([USE_COLLECTIONS]);
```
Comment (tihuan) [inline frontend/src/common/queries/collections.ts:197]: Looks like `invalidateQueries()` now expects an array

File: frontend/src/common/queries/README.md [hunk 5]
```diff
@@ -60,9 +60,9 @@ The benefits of this arrangement include:
    code to achieve that:
 
    ```ts
-   import { useMutation, useQueryCache } from "react-query";
+   import { useMutation, useQueryClient } from "react-query";
 
-   const queryCache = useQueryCache();
+   const queryClient = useQueryClient();
```
Comment (tihuan) [inline frontend/src/common/queries/README.md:65]: https://react-query.tanstack.com/guides/migrating-to-react-query-3#the-querycache-has-been-split-into-a-queryclient-and-lower-level-querycache-and-mutationcache-instances

File: frontend/src/components/Collections/components/DeleteCollection/index.tsx [hunk 6]
```diff
@@ -22,7 +22,7 @@ const DeleteCollection: FC<Props> = ({
   Button = RawButton,
   isRevision,
 }) => {
-  const [deleteMutation, { isLoading }] = useDeleteCollection(id);
+  const { mutateAsync: deleteMutation, isLoading } = useDeleteCollection(id);
```
Comment (tihuan) [inline frontend/src/components/Collections/components/DeleteCollection/index.tsx:25]: https://react-query.tanstack.com/guides/migrating-to-react-query-3#mutationmutate-no-longer-return-a-promise

File: frontend/src/components/Collections/components/DeleteCollection/index.tsx [hunk 7]
```diff
@@ -22,14 +22,13 @@ const DeleteCollection: FC<Props> = ({
...
   const router = useRouter();
 
   const handleDelete = async () => {
-    await deleteMutation(
-      { collectionID: id },
-      { onSuccess: () => router.push(ROUTES.MY_COLLECTIONS) }
```
Comment (tihuan) [inline frontend/src/components/Collections/components/DeleteCollection/index.tsx:31]: pretty crazy, it seems like `mutateAsync` is not invoking `onSuccess()` callback for whatever reason, even though it should support callbacks. So to get around the issue, I'm just awaiting `deleteMutation()` first, then push route!

Comment (seve) [reply frontend/src/components/Collections/components/DeleteCollection/index.tsx:31]: Could we add that as a comment?

**Review by seve — APPROVED** [review]
just one comment!

File: frontend/src/components/Collections/components/DeleteCollection/index.tsx (hunk 7 above)
Comment (tihuan) [reply frontend/src/components/Collections/components/DeleteCollection/index.tsx:31]: Oh yes great idea! Done!

**Pull request merged**